google/cloud/spanner/internal/time_format.cc

```diff
@@ -34,6 +37,62 @@ namespace spanner {
 inline namespace SPANNER_CLIENT_NS {
 namespace internal {
 
+namespace {
+
+constexpr auto kDigits = "0123456789";
+
+// Formats [0 .. 99] as %02d.
+inline char* Format02d(char* ep, int v) {
+  *--ep = kDigits[v % 10];
+  *--ep = kDigits[(v / 10) % 10];
+  return ep;
+}
+
+char const* ParseInt(char const* dp, int min, int max, int* vp) {
+  int value = 0;
+  bool neg = false;
+  if (*dp == '-') {
+    neg = true;
+    ++dp;
+  }
+  char const* const bp = dp;
+  constexpr int kMin = std::numeric_limits<int>::min();
+  while (char const* cp = std::strchr(kDigits, *dp)) {
+    int d = static_cast<int>(cp - kDigits);
+    if (d >= 10) break;  // non-digit
+    if (value < kMin / 10) return nullptr;
+    value *= 10;
+    if (value < kMin + d) return nullptr;
+    value -= d;
+    ++dp;
+  }
+  if (dp == bp) return nullptr;
+  if (!neg) {
+    if (value == kMin) return nullptr;
+    value = -value;  // make positive
+  }
+  if (value < min || max < value) return nullptr;
+  *vp = value;
+  return dp;
+}
+
+inline bool LeapYear(int y) {
+  return y % 4 == 0 && (y % 100 != 0 || y % 400 == 0);
+}
+
+// Note: tm.tm_year and tm.tm_mon are unadjusted (i.e., have true values).
+bool ValidDay(std::tm const& tm) {
+  static constexpr std::array<int, 1 + 12> kMonthDays = {
+      {-1, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}  // non leap year
+  };
+  if (tm.tm_mon == 2 && LeapYear(tm.tm_year)) {
+    return tm.tm_mday <= 29;
+  }
+  return tm.tm_mday <= kMonthDays[tm.tm_mon];
```

Comment:
> Should we worry about out of range values for tm.tm_mon?

No ... we only pass a `tm` whose `tm_mon` and `tm_day` fields have already been range-checked to [1:12] and [1:31] respectively.